Mathlib/GroupTheory/Perm/Cycle/Basic.lean

```diff
@@ -1978,3 +2084,210 @@ theorem _root_.Finset.sum_mul_sum_eq_sum_perm (hσ : σ.IsCycleOn s) (f g : ι 
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ g x ∈ c.support,
+        ofSubtype (subtypePerm g hc') ∈ Subgroup.zpowers c := by
+  rw [IsCycle.commute_iff' hc]
+  apply exists_congr
+  intro hc'
+  simp only [Subgroup.mem_zpowers_iff]
+  apply exists_congr
+  intro k
+  unfold subtypePermOfSupport
+  rw [subtypePerm_zpow]
+  simp only [Equiv.ext_iff, subtypePerm_apply, Subtype.mk.injEq, Subtype.forall]
+  apply forall_congr'
+  intro a
+  by_cases ha : a ∈ c.support
+  · rw [imp_iff_right ha]
+    apply Eq.congr rfl
+    apply symm
+    exact ofSubtype_apply_of_mem _ (by exact ha)
+  · rw [iff_true_left _]
+    · rw [ofSubtype_apply_of_not_mem, ← not_mem_support]
+      · exact Finset.not_mem_mono (support_zpow_le c k) ha
+      exact ha
+    exact fun b ↦ (ha b).elim
+
+theorem IsCycle.forall_commute_iff (g z : Perm α) :
+    (∀ c ∈ g.cycleFactorsFinset, Commute z c) ↔
+      ∀ c ∈ g.cycleFactorsFinset,
+      ∃ (hc : ∀ x : α, x ∈ c.support ↔ z x ∈ c.support),
+        ofSubtype (subtypePerm z hc) ∈ Subgroup.zpowers c := by
+  apply forall_congr'
+  intro c
+  apply imp_congr_right
+  intro hc
+  exact IsCycle.commute_iff (mem_cycleFactorsFinset_iff.mp hc).1
+
+/-- A permutation restricted to the support of a cycle factor is that cycle factor -/
+theorem subtypePerm_on_cycleFactorsFinset {g c : Perm α}
+    (hc : c ∈ g.cycleFactorsFinset) :
+    g.subtypePerm (mem_cycleFactorsFinset_support hc) = c.subtypePermOfSupport := by
+  ext ⟨x, hx⟩
+  simp only [subtypePerm_apply, Subtype.coe_mk, subtypePermOfSupport]
+  exact ((mem_cycleFactorsFinset_iff.mp hc).2 x hx).symm
+
+theorem commute_of_mem_cycleFactorsFinset_iff {g k c : Equiv.Perm α}
+    (hc : c ∈ g.cycleFactorsFinset) :
+    Commute k c ↔
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ k x ∈ c.support,
+        k.subtypePerm hc' ∈ Subgroup.zpowers
+          (g.subtypePerm (mem_cycleFactorsFinset_support hc)) := by
+  rw [IsCycle.commute_iff' (mem_cycleFactorsFinset_iff.mp hc).1]
+  apply exists_congr
+  intro hc'
+  simp only [Subgroup.mem_zpowers_iff]
+  apply exists_congr
+  intro n
+  unfold subtypePermOfSupport
+  rw [Equiv.Perm.subtypePerm_on_cycleFactorsFinset hc]
+  rfl
+
+theorem zpow_eq_ofSubtype_subtypePerm_iff
+    {g c : Equiv.Perm α} {s : Finset α}
+    (hg : ∀ x, x ∈ s ↔ g x ∈ s) (hc : c.support ⊆ s) (n : ℤ) :
+    c ^ n = ofSubtype (g.subtypePerm hg) ↔
+      c.subtypePerm (isInvariant_of_support_le hc) ^ n = g.subtypePerm hg := by
+  constructor
+  · intro h; ext ⟨x, hx⟩; let h' := Perm.congr_fun h x
+    simp only [h', subtypePerm_apply_zpow_of_mem, Subtype.coe_mk, subtypePerm_apply]
+    rw [ofSubtype_apply_of_mem]
+    simp only [Subtype.coe_mk, subtypePerm_apply]
+    exact hx
+  · intro h; ext x
+    rw [← h]
+    by_cases hx : x ∈ s
+    · rw [ofSubtype_apply_of_mem]
+      simp only [subtypePerm_zpow, subtypePerm_apply]
+      exact hx
+    · rw [ofSubtype_apply_of_not_mem, ← not_mem_support]
+      exact fun hx' ↦ hx (hc (support_zpow_le _ _ hx'))
+      exact hx
```

Comment:
```suggestion
    · simp only [Subtype.coe_mk, subtypePerm_apply]
    · exact hx
  · intro h; ext x
    rw [← h]
    by_cases hx : x ∈ s
    · rw [ofSubtype_apply_of_mem]
      · simp only [subtypePerm_zpow, subtypePerm_apply]
      · exact hx
    · rw [ofSubtype_apply_of_not_mem, ← not_mem_support]
      · exact fun hx' ↦ hx (hc (support_zpow_le _ _ hx'))
      · exact hx
```

It's annoying you can't pass `hx` to `ofSubtype_apply_of_{not_,}mem` directly. Have you looked into why that is?

Reply:
One can, but one needs to fill it a big share of the `subtypePerm`